src/ir/auth_logic/ast.h

```diff
@@ -75,24 +86,34 @@ class Attribute {
     return visitor.Visit(*this);
   }
 
-  // A potentially ugly print of the state in this class
-  // for debugging/testing only
-  std::string DebugPrint() const {
-    return absl::StrCat(principal_.name(), predicate_.DebugPrint());
+  bool operator==(const Attribute& rhs) const {
+    return principal_ == rhs.principal() && predicate_ == rhs.predicate();
   }
 
+  virtual bool operator==(const ASTNode& rhs) const override {
+    try {
+      const Attribute & rhsAttribute = dynamic_cast<const Attribute&>(rhs);
+      return *this == rhsAttribute;
+    } catch(const std::bad_cast& e) {
+      return false;
+    }
```

Comment:
Rather than having this `try..catch` structure, just `dynamic_cast` to `const Attribute *` like so:

```C++
const Attribute *rhs_attribute = dynamic_cast<const Attribute *>(&rhs);
```

If the cast fails, `rhs` will contain `nullptr`, and you can return false in that case, like so:

```C++
if (const Attribute *rhs_attribute = dynamic_cast<const Attribute *>(&rhs)) {
  return *this == *rhs_attribute;
}
return false;
```

Reply:
fixed